Rework updatemanager

niklasweimann opened the pull request:
Special Thanks to: https://github.com/HedgehogNSK

The main purpose of this PR is to add the ability to switch between different update streams. It allows users to switch between various streams, such as long polling, webhooks, or test streams, without needing to resubscribe to a new update manager.

Core changes:

IUpdateManager extended with the Set(IObservable<Update>) method.
UpdateManager has been split into two classes: UpdateDistributor and LongPollingUpdateTracker.
UpdateDistributor serves as a facade that implements IUpdateManager.
LongPollingUpdateTracker implements IObservable<Update>.
UpdateManager is no longer needed but has been retained as is.
The TelegramBot class has been extended with a Builder that allows setting IUpdateManager and IObservable<Update>.

HedgehogNSK committed:
Split UpdateManager onto UpdateDistribtor and UpdateTracker entities.
UpdateDistributor implements IUpdateManager
UpdateTracker implements IObservable<Update>
Add TelegramBot.Builder
Extend IUpdateManager with Set(IObservable<Update>) method
Add Rx utility entities: CustomSubjects that works similar to Rx Subject
DisposableAction that  works as Rx Disposable.Create

HedgehogNSK committed:
Fix TelegramBot.Builder.Build()
Desc: It didn't apply tracker to update manager

HedgehogNSK committed:
Add TelegramBot.Builder.SetToken(string)

niklasweimann committed:
Merge pull request #70 from HedgehogNSK/feature/rework_updatemanager

Feature/rework updatemanager

HedgehogNSK commented:
Hi, there. 😃  
Should I fix the issues that Sonar is flagging?

HedgehogNSK committed:
Fix additional subscription to same UpdateType

HedgehogNSK committed:
Apply Sonar recommendations:
- Move UpdateDistributor._update initialization to constructor
- Add 'readonly' keyword to LongpollingUpdateTracker._observers
- Add 'readonly' keyword to UpdateDistributor._updateInfos
- Add 'sealed' keyword to UpdateTypeInfo
- Add 'sealed' keyword to DisposableAction
- Call Dispose method for LongpollingUpdateTracker._cancellationTokenSource
- Add class constraint to TOut parameter in CustomSubject class

HedgehogNSK commented:
Please check PR #74. Fixed a bug and addressed some issues pointed out by Sonar.

HedgehogNSK committed:
Rework `UpdateDistributor` tracking mechanism and fix exception handling issues.

- Added Observable operators:
  - `DoOnDispose`: Executes an action when the observable is disposed.
  - `DoOnSubscribe`: Executes an action when a new observer subscribes.
  - `Select`: Applies a function to each item in the observable and returns the result of that function.
  - `Switch`: Switches to the latest observable sequence when the source emits a new value.
  - `Where`: Filters the observable sequence based on a predicate.

- Added `ReactiveProperty` for easier state management.
- Removed unnecessary `CustomSubject` and `UpdateSubject`.
- Fixed exception handling in `LongpollingUpdateTracker`.

HedgehogNSK committed:
Remove unnecessary comments in `UpdateDistributor`

HedgehogNSK committed:
Remove unnecessary logs in `UpdateDistributor`

HedgehogNSK committed:
Fixed update type assignment order for tracker

HedgehogNSK committed:
Added `Finally` observable operator.
Replaced `DoOnDispose` with `Finally` in `UpdateDistributor` to fix subscriber tracking.

niklasweimann committed:
Merge pull request #74 from HedgehogNSK/feature/rework_updatemanager

Feature/rework updatemanager

niklasweimann commented:
@HedgehogNSK Two Issues here:
1. Could you please fix the 9 issues that sonar cloud found?
2. Cloud you please make sure that your Source Code matches the rest of the project? We use [Rider](https://www.jetbrains.com/rider/download/#section=mac) for the development so when you use it too you should be able to see the issues. Otherwise please make sure your editor respects the .editorconfig of the project (I have not tested it with other IDEs but it should result in a similar style).

HedgehogNSK commented:
Ok. :) I need time a bit time and I will fix issues. ;)

niklasweimann commented:
Awesome 👍😎

HedgehogNSK committed:
Apply .editorconfig

niklasweimann commented:
Hey @HedgehogNSK,

The following implementation should meet the sonar cloud requirements:
``` sharp
~ReactiveProperty() => Dispose(false);

    public void Dispose()
    {
        Dispose(true);
        GC.SuppressFinalize(this);
    }

    protected virtual void Dispose(bool explicitDisposing)
    {
        if (IsDisposed)
        {
            return;
        }

        if (explicitDisposing)
        {
            OnCompleted();
            Interlocked.Exchange(ref _observers, Terminated);
        }

        IsDisposed = true;
    }
```

When looking at the SonarCloud issue you can click on "Why is this an issue?" and near the end is an example of how to implement disposable properly.

HedgehogNSK committed:
Fix Sonar issues

niklasweimann committed:
Merge pull request #75 from HedgehogNSK/feature/rework_updatemanager

Feature/rework updatemanager Fixes for Sonar and Config issues for PR #71

HedgehogNSK committed:
Sonar fixes:
- Fix IDisposable for ReactiveProperty
- Fix IDisposable for SwithObservable

niklasweimann committed:
Merge pull request #76 from HedgehogNSK/feature/rework_updatemanager
